src/Pirouette/Symbolic/Eval.hs

```diff
@@ -216,25 +219,34 @@ instance (SymEvalConstr lang) => Alternative (SymEval lang) where
               (runSymEvalRaw env st xs, runSymEvalRaw env st ys)
        in xs' <|> ys'
 
--- | Learn a new constraint and add it as a conjunct to the set of constraints of
---  the current path, then checks whether it leads to an inconsistency and fails
--- if it does.
-learn :: (SymEvalConstr lang) => [C.Constraint lang SymVar] -> SymEval lang ()
-learn cs = do
+pruneOrGetPathProblem :: (SymEvalConstr lang) => [C.Constraint lang SymVar] -> SymEval lang (SolverProblem lang Bool)
+pruneOrGetPathProblem cs = do
   curr <- gets sestConstraint
-  -- Add the conjunct. `C.conjunct` fails when it detects trivial inconsistencies.
   case foldl' (\mc c -> mc >>= C.conjunct c) (Just curr) cs of
     Nothing -> empty
-    Just curr' ->
-      -- As far as `C.conjunct` is concerned, there is no inconsistencies. Let us
-      -- call the solver to actually check that.
-      SymEval $
-        ReaderT $ \env -> do
-          st <- get
-          let st' = st {sestConstraint = curr'}
-          if solvePathProblem (seeSolvers env) (CheckPathProblem st' (seeDefs env))
-            then put st'
-            else empty
+    Just curr' -> do
+      st <- get
+      defs <- asks seeDefs
+      let st' = st {sestConstraint = curr'}
+      put st'
+      return $ CheckPath (CheckPathProblem st' defs)
+
+solveAll :: forall lang res t. Traversable t => SymEval lang (t (SolverProblem lang res)) -> SymEval lang (t res)
+solveAll (SymEval f) = SymEval $
+  ReaderT $ \env -> StateT $ \st -> ListT.Action @IO $ do
+    problems <- ListT.runActions $ runStateT (runReaderT f env) st
+    let -- change the type to get a Traversable instance
+        problemsTraversable = Compose $ Compose . swap <$> problems
+    -- solve the problems asynchronously
+    resultsTraversable <- seeAsyncSolver env @(Compose WeightedList (Compose ((,) (SymEvalSt lang)) t)) problemsTraversable
+    let -- get the correct type back
+        results = swap . getCompose <$> getCompose resultsTraversable
+    return $ ListT.mapWeightedListT (pure . runIdentity) results
```

Comment:
Informally, `solveAll`
- takes a `WeightedList` of pairs of `t (SolverProblem lang res)` and `SymEvalSt lang`, 
- transforms it so that the overall functor wrapping the problem type `WeightedList o (,) (SymEvalSt lang) o t`) is recognized as `Traversable`, 
- solves the problems asynchronously, 
- transform the results back into a `WeightedList`
- make the `WeightedList` into a `WeightedListT IO` so we can wrap the whole computation inside the `SymEval` monad.